src/CoreWCF.Primitives/src/CoreWCF/Telemetry/PropertyFetcher.cs

```diff
@@ -0,0 +1,123 @@
+// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+
+using System;
+using System.Linq;
+using System.Reflection;
+
+namespace CoreWCF.Telemetry;
+
+/// <summary>
+/// PropertyFetcher fetches a property from an object.
+/// </summary>
+/// <typeparam name="T">The type of the property being fetched.</typeparam>
+#pragma warning disable CA1812
+internal sealed class PropertyFetcher<T>
+#pragma warning restore CA1812
+{
+    private readonly string propertyName;
+    private PropertyFetch innerFetcher;
+
+    /// <summary>
+    /// Initializes a new instance of the <see cref="PropertyFetcher{T}"/> class.
+    /// </summary>
+    /// <param name="propertyName">Property name to fetch.</param>
+    public PropertyFetcher(string propertyName)
+    {
+        this.propertyName = propertyName;
+    }
+
+    /// <summary>
+    /// Fetch the property from the object.
+    /// </summary>
+    /// <param name="obj">Object to be fetched.</param>
+    /// <returns>Property fetched.</returns>
+    public T Fetch(object obj)
+    {
+        return !this.TryFetch(obj, out var value)
+            ? throw new ArgumentException("Supplied object was null or did not match the expected type.", nameof(obj))
+            : value;
+    }
+
+    /// <summary>
+    /// Try to fetch the property from the object.
+    /// </summary>
+    /// <param name="obj">Object to be fetched.</param>
+    /// <param name="value">Fetched value.</param>
+    /// <returns><see langword="true"/> if the property was fetched.</returns>
+    public bool TryFetch(object obj, out T value)
```

Comment:
I understand what's trying to be done here. There's some problems with the code, eg there's no check that the inner fetcher is for the same type as obj. I think we might be able to avoid the need for this though. We add the HttpContext as a message property (we use the fully qualified class name as the key), and that is a resolvable type in Primitives. You can read the headers from there without dealing with any reflection.